Update to GraalVM / Mandrel 22.3.0

cescoffier opened the pull request:
DO NOT MERGE!

cescoffier committed:
Update GraalVM 22.3.0 images

zakkak commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -656,7 +656,7 @@ Sample multistage Dockerfile for building an image from `scratch`:
 ----
 ## Stage 1 : build with maven builder image with native capabilities
-FROM quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.2-java11 AS build
+FROM quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.3-java11 AS build
```
This should be:

```suggestion
FROM quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:{graalvm-flavor} AS build
```

zakkak reviewed (dismissed):
Needs rebasing now that https://github.com/quarkusio/quarkus/pull/28843 got merged.

gsmet commented:
Let me rebase it.

gsmet commented:
Done! We still need to adjust the Mandrel image version when available but I updated the doc version as it won't break anything.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:
```diff
@@ -16,7 +16,7 @@
 
-    public static final String DEFAULT_GRAALVM_BUILDER_IMAGE = "quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.2-java17";
+    public static final String DEFAULT_GRAALVM_BUILDER_IMAGE = "quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.3-java17";
     public static final String DEFAULT_MANDREL_BUILDER_IMAGE = "quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.2-java17";
```
We need to update that one when ready.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:
```diff
@@ -16,7 +16,7 @@
 
-    public static final String DEFAULT_GRAALVM_BUILDER_IMAGE = "quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.2-java17";
+    public static final String DEFAULT_GRAALVM_BUILDER_IMAGE = "quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.3-java17";
     public static final String DEFAULT_MANDREL_BUILDER_IMAGE = "quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.2-java17";
```
To make sure we won't forget: we need to update that one when ready.

jerboaa replied:
`quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.3-java17` is there now, fwiw.

gsmet commented:
@zakkak I see the build failing with:
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project quarkus-core: Compilation failure: Compilation failure: 
Error:  /home/runner/work/quarkus/quarkus/core/runtime/src/main/java/io/quarkus/runtime/graal/ConfigurationSubstitutions.java:[6,36] cannot find symbol
Error:    symbol:   class AlwaysInline
Error:    location: package com.oracle.svm.core.annotate
Error:  /home/runner/work/quarkus/quarkus/core/runtime/src/main/java/io/quarkus/runtime/graal/ConfigurationSubstitutions.java:[26,6] cannot find symbol
Error:    symbol:   class AlwaysInline
Error:    location: class io.quarkus.runtime.graal.Target_io_smallrye_config_SmallRyeConfigProviderResolver
Error:  -> [Help 1]
```

Did we miss something when ensuring `22.3` compatibility?

zakkak commented:
> Did we miss something when ensuring `22.3` compatibility?

Hmmm, indeed. Unfortunately Quarkus uses the defined `graalvm-version` for the GraalVM artifacts no matter which GraalVM version we use. So in our tests javac was building with the 22.2 GraalVM jars despite using 22.3 to build the actual native image. IIRC @Sanne experienced this at some point as well.

Regarding this specific failure, it's because [`com.oracle.svm.core.annotate.AlwaysInline` (which is not public API) was moved to `com.oracle.svm.core.AlwaysInline`](https://github.com/oracle/graal/pull/4683/files#diff-e9ce1eb4db69a014a3366a094cffd77696668c25de9094c6628168a0e9193bda).

So it looks like we will need to patch the source code and bump the minimum version to 22.3.

gsmet committed:
Use the GraalVM 22.3 AlwaysInline annotation

This annotation has moved to a new package and we need to use the 22.3
version to make sure our code compiles.
As for using 22.2, it still works but the annotation is ignored, which
looks OK as the annotation is not used in places where it is strictly
necessary.
If people use the recommended GraalVM version, everything will be fine.
If not, things will mostly work anyway.

gsmet commented:
I pushed an update to move to the new package. Rationale explained in the commit comment.

cescoffier commented:
@zakkak unfortunately, it cannot be backported, at least not directly.
These images use the new multi-archs and jdock models. We need to use the former model for the backport.

zakkak commented:
> @zakkak unfortunately, it cannot be backported, at least not directly. These images use the new multi-archs and jdock models. We need to use the former model for the backport.

Oh, I see. Shall we keep the label though to indicate that we need this in 2.13 as well?

gsmet commented:
Yes, let's keep the label so that we don't miss it.

gsmet commented:
@cescoffier what's the status of this build in your fork? Did you see other problems?

cescoffier commented:
Nope, looks good.
The 2.13 images have been produced. Waiting for mandrel.

zakkak commented:
> Waiting for mandrel.

FYI, due to some infrastructure related issues this might take a bit longer than usual.

zakkak commented:
Mandrel 22.3 is now released. I have also openned https://github.com/quarkusio/quarkus-images/pull/215 to generate the quarkus images. @cescoffier are you going to update this PR to include the Mandrel images as well?

cescoffier committed:
Update to Mandrel 22.3 images

zakkak reviewed (approved):
LGTM, thanks @cescoffier

gsmet commented:
I rebased to trigger CI as it was in limbo.

rsvoboda commented:
I see errors in our daily runs, it's apparently because of https://github.com/oracle/graal/issues/5303 bug

Shouldn't we set mandrel image as default for now? Peter has commented on the issue that Mandrel 22.3 works for him.

zakkak commented:
@gsmet I am marking this as a "noteworthy" feature for better visibility.

(inspired by discussion in https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Quarkus.20won't.20build.20native/near/307716675)

zakkak commented:
@rsvoboda I really don't know. I personally have not seen this issue happen in any of the machines I have access to.

I would suggest commenting on the upstream issue and possibly providing more info about the context under which the issue is triggered.

Regarding switching to Mandrel by default I am +0. People getting hit by the issue can easily switch to Mandrel by passing `-Dquarkus.native.builder-image=mandrel`. If it turns out to be a pretty common issue I expect upstream to make a new release soonish.

rsvoboda commented:
ok, I will comment upstream
